### PR TITLE
WIP: DNA: Warn when using Jetpack Packages before `plugins_loaded`

### PR DESCRIPTION
**Pull request by lezama**
Waiting for `plugins_loaded` is the right thing to do if we want the site to always load the latest available version of a JP Package.
As Today Jetpack does a lot of invocations before `plugins_loaded`, throwing Notices when `defined( 'WP_DEBUG' ) && WP_DEBUG` for the following classes:

```
Automattic\Jetpack\Connection\Manager
Automattic\Jetpack\Connection\Manager_Interface
Jetpack_Options
Jetpack_Sync_Main
Automattic\Jetpack\Constants
```

Open questions:

- why would we want to do anything before `plugins_loaded`?
- either do it after `plugins_loaded` or live with the first found version of a class?
- HOW SHOULD WE MOVE FORWARD?

**Testing Instructions**
* run `composer dump-autoload` to generate a new `autoload_packages.php`
* make sure `WP_DEBUG` is set to true
* after loading the site confirm there are no warnings in the frontend or in the logs
* comment one of the classes in the '$ignore` array
* after loading the site confirm there are warnings in the frontend or in the logs

### CODE REVIEW

File: packages/autoloader/src/autoload.php [hunk 1]
```diff
@@ -73,7 +73,7 @@ function autoloader( $class_name ) {
 
 		if ( isset( $jetpack_packages_classes[ $class_name ] ) ) {
 
-			if ( defined( WP_DEBUG ) && WP_DEBUG ) {
+			if ( defined( 'WP_DEBUG' ) && WP_DEBUG ) {
```
Comment (tyxla) [inline packages/autoloader/src/autoload.php]: How did this even work 😱

File: packages/autoloader/src/autoload.php [hunk 2]
```diff
@@ -94,6 +94,7 @@ function autoloader( $class_name ) {
 				return true;
 			}
 		}
+		return false;
```
Comment (tyxla) [inline packages/autoloader/src/autoload.php]: Why do we need this? For consistent method signature, or?

Comment (tyxla) [reply packages/autoloader/src/autoload.php]: I'm asking because the default `spl_autoload` doesn't return anything AFAIK.

Comment (lezama) [reply packages/autoloader/src/autoload.php]: yeah, phpcs was mild complaining

**Review by tyxla — COMMENTED** [review]
I'd say that if we have to do things before `plugins_loaded`, we'll want to be graceful. So, if we need to autoload a class before the action has fired, we could autoload what we already have (use the default autoloader). And if we need to autoload a class after the action has fired, we can use the custom autoloader.

The downside with this is that we'll lose the full control over what versions of the packages we actually load.

File: packages/autoloader/src/autoload.php (hunk 2 above)
Comment (tyxla) [reply packages/autoloader/src/autoload.php]: Could we `void` it?

Comment (lezama) [reply packages/autoloader/src/autoload.php]: it needs to return `true` when it finds the class. so autoloading stops searching

File: packages/autoloader/src/autoload.php [hunk 3]
```diff
@@ -72,9 +72,20 @@ function autoloader( $class_name ) {
...
+				$ignore = in_array(
+					$class_name,
+					array(
+						'Automattic\Jetpack\JITM\Manager',
+						'Automattic\Jetpack\Connection\Manager',
+						'Automattic\Jetpack\Connection\Manager_Interface',
+						'Jetpack_Options',
+						'Jetpack_Sync_Main',
+						'Automattic\Jetpack\Constants',
+					)
```
Comment (tyxla) [inline packages/autoloader/src/autoload.php]: Let's use strict `in_array`.

```suggestion
					),
					true
```

**Review by tyxla — DISMISSED** [review]
This tests well 👍 Left a small suggestion.

I like the idea for having exceptions. Let's just try to keep them to a bare minimum, ideally we'd end up having none.

File: packages/autoloader/src/autoload.php [hunk 4]
```diff
@@ -72,9 +72,21 @@ function autoloader( $class_name ) {
...
 		if ( isset( $jetpack_packages_classes[ $class_name ] ) ) {
-
-			if ( defined( WP_DEBUG ) && WP_DEBUG ) {
-				if ( function_exists( 'did_action' ) && ! did_action( 'plugins_loaded' ) ) {
+			if ( defined( 'WP_DEBUG' ) && WP_DEBUG ) {
+				// TODO ideally we shouldn't skip any of these, see: https://github.com/Automattic/jetpack/pull/12646
+				$ignore = in_array(
+					$class_name,
+					array(
+						'Automattic\Jetpack\JITM\Manager',
```
Comment (tyxla) [inline packages/autoloader/src/autoload.php]: We just changed this one:

```suggestion
						'Automattic\Jetpack\JITM',
```

**Review by tyxla — APPROVED** [review]
Tests well for me 👍 Let's 🚢 and iterate.

**Review by dereksmart — APPROVED** [review]

**Pull request merged**